src/core/device.c

```diff
@@ -891,7 +891,7 @@ static int device_dispatch_io(sd_device_monitor *monitor, sd_device *dev, void *
         /* A change event can signal that a device is becoming ready, in particular if
          * the device is using the SYSTEMD_READY logic in udev
          * so we need to reach the else block of the follwing if, even for change events */
-        if (streq(action, "remove"))  {
+        if (STR_IN_SET(action, "remove", "unbind"))  {
                 r = swap_process_device_remove(m, dev);
                 if (r < 0)
                         log_device_warning_errno(dev, r, "Failed to process swap device remove event, ignoring: %m");
```

Comment:
This part seems dubious. Calling `swap_process_device_remove` seems right, since we don't want to have swap on an unbound device. But the second part, unconditionally calling `device_update_found_by_sysfs(..., 0, ...)` might be wrong. After all, the device *is* in sysfs and the user might even have rules that do `TAGS+=systemd` on it.

Reply:
> This part seems dubious. Calling `swap_process_device_remove` seems right, since we don't want to have swap on an unbound device.

No, we don't care / it never happens. This is the confusion that I was worried this patch would propagate.  Hence my hope we can document it somewhere, if it is too complex to put a comment on each piece of confusing code.

You should never get an unbind event on a block device.  You never bind drivers to block devices.  Block devices are class devices, therefore they are not bus devices and do not have drivers.  There are some character devices which are bus devices, e.g. `/dev/bus/usb/...`.  But you can't add a character device as a swap device.

Most of the devices we care about are class devices.  I think the most common reason to worry about bus devices is around userspace USB stuff.

(What does `unbind` of the `usb` driver from the usb character device do?  Well, all the child usb devices disappear.  And the chardev ioctls that work by binding the `usbfs` driver to a specific child device - I *think* those are aborted with `ENOENT`.  But the character device is still there, and there's a "reset usb device" ioctl that doesn't reference children so I expect that to still work, for example.  But I haven't found any practical use for unbinding the `usb` driver - no alternative driver that you can bind to these devices.

In practice, our code is only having a problem because `unbind` is being emitted during `remove`, and *only* because that triggers a specific race condition in udev software.  It would be real nice if udev didn't have any race conditions that left phantom devices around when you get change remove event before we've started processing the last change event.  The kernel triggering this race every time is arguably an ABI break, but of course it wasn't noticed or reported to them at the time, so now we have 12+ months of kernel versions with the new behaviour.  (The followup patch for the MODALIAS issue is a different issue)).

Reply:
> FWIW, I can recreate the reported problem fairly easily with manual unbind:
> 
> ```
> $ cat /etc/udev/rules.d/90-yubikey-debug.rules
> ACTION=="add|change|move", ENV{ID_VENDOR}=="Yubico", TAG+="systemd"
> $ sudo bash -c 'echo -n 1-4.4 >/sys/devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4.4/driver/unbind'
> $ systemctl status /dev/bus/usb/001/015
> ● dev-bus-usb-001-015.device - Yubikey 4 OTP+U2F+CCID
>    Follow: unit currently follows state of sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d4-1\x2d4.4.device
>    Loaded: loaded
>    Active: active (plugged) since Thu 2018-12-06 11:48:12 CET; 41min ago
>    Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4.4
> ```

Not correct.

I can reproduce the problem with a USB printer.  The problem shows as a device which stays as unplug.  So if I plug and unplug the device three times, I get three devices hanging around which should not be there.

You will find that if you do manual unbind+bind several times, you do not accumulate extra phantom devices.